### PR TITLE
* NEW: Add `--disable-telemetry` argument to disable telemetry collection

### PR DESCRIPTION
**Pull request by shaopeng-gh**
If you have the telemetry connection string set, 
you can use this new setting to quickly disable it without actually remove the setting,
so that your detail scan results will not be send.

### CODE REVIEW

File: src/BinSkim.Sdk/EnvironmentResources.resx [hunk 1]
```diff
@@ -0,0 +1,123 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (scottoneil-ms) [inline src/BinSkim.Sdk/EnvironmentResources.resx:1]: I do not know how to review this, you might need a second set of eyes for it.

File: src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs [hunk 2]
```diff
@@ -172,6 +173,11 @@ public override int Run(AnalyzeOptions analyzeOptions)
         {
             Stopwatch stopwatch = null;
 
+            if (analyzeOptions.DisableTelemetry == true)
+            {
+                this.Telemetry = null;
```
Comment (scottoneil-ms) [inline src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs:178]: This worries me a bit.  I see that if it's null, we just don't add a certain logger to an aggregate logger.  That looks great.

But then a little bit later, we pass this thing into a caller.  Does that call flow lead to uses?  Is it already nullchecked?

File: src/Test.FunctionalTests.BinSkim.Driver/CommandLineTests.cs [hunk 3]
```diff
@@ -70,5 +70,40 @@ public void MostlyFunctionlessCommandlineTest()
             builder.Length.Should().Be(0,
                 $"all test cases should pass, but the following test cases failed:\n{builder}");
         }
+
+        [Fact]
+        public void DisableTelemetryCommandlineTest()
```
Comment (scottoneil-ms) [inline src/Test.FunctionalTests.BinSkim.Driver/CommandLineTests.cs:75]: I don't know what the rest of this suite looks like, and I'm not 100% sure I'm following *this* test.   So tell me if this is here and I'm not seeing it.  But: I think you need a test that disables telemetry, runs a scan against a well-known file (I see this much) and then verifies that the scan results are the same as if you had not passed in disable-telemetry (I don't see this part.)

File: src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs (hunk 2 above)
Comment (shaopeng-gh) [reply src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs:178]: yes it is, 
in fact our application is checking if it is null to see enabled or not, 
if null, no Telemetry will be performed.

File: src/Test.FunctionalTests.BinSkim.Driver/CommandLineTests.cs (hunk 3 above)
Comment (shaopeng-gh) [reply src/Test.FunctionalTests.BinSkim.Driver/CommandLineTests.cs:75]: good input. Added another test.

File: src/BinSkim.Sdk/EnvironmentResources.resx (hunk 1 above)
Comment (shaopeng-gh) [reply src/BinSkim.Sdk/EnvironmentResources.resx:1]: the .net resource file is auto gen,
the only thing that is worth looking is the `DefaultTelemetryConnectionString` at the end.

**Comment by stacywray** [comment]
@shaopeng-gh Release history looks good.

**Review by stacywray — APPROVED** [review]
:shipit:

File: src/Test.FunctionalTests.BinSkim.Driver/CommandLineTests.cs (hunk 3 above)
Comment (michaelcfanning) [inline src/Test.FunctionalTests.BinSkim.Driver/CommandLineTests.cs:75]: >DisableTelemetryCommandlineTest [](http://example.com/codeflow?start=20&length=31)

i don't see any tests that prove that telemetry isn't collected in a tool run where this setting is specified. how do we know it works?

File: src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs:178]: Scott's right, you haven't gotten this change quite right, which is to *enable* telemetry collection by default but allow someone to override that via this argument.

Where is the code that makes telemetry work out of box with no special configuration?

**Review by michaelcfanning — APPROVED** [review]
:shipit:

**Pull request merged**